internal/providers/providers.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/K0rdent/kcm/internal/credspropagation"
+)
+
+const (
+	// InfraPrefix is the prefix used for infrastructure provider names
+	InfraPrefix = "infrastructure-"
+	// ProviderPrefix is the prefix used for cluster API provider names
+	ProviderPrefix = "cluster-api-provider-"
+)
+
+var (
+	mu sync.RWMutex
+
+	providers = []hmc.Provider{
+		{
+			Name: hmc.ProviderK0smotronName,
+		},
+		{
+			Name: hmc.ProviderSveltosName,
+		},
+	}
+
+	registry map[string]ProviderModule
+)
+
+type ProviderModule interface {
+	// GetName returns the short name of the provider
+	GetName() string
+	// GetTitleName returns the display title of the provider
+	GetTitleName() string
+	// GetClusterGVK returns the GroupVersionKind for the provider's cluster resource
+	GetClusterGVK() schema.GroupVersionKind
+	// GetClusterIdentityKinds returns a list of supported cluster identity kinds
+	GetClusterIdentityKinds() []string
+	// CredentialPropagationFunc returns a function to handle credential propagation
+	CredentialPropagationFunc() func(
+		ctx context.Context,
+		cfg *credspropagation.PropagationCfg,
+		l logr.Logger,
+	) (enabled bool, err error)
+}
+
+// Register adds a new provider module to the registry
+func Register(p ProviderModule) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if registry == nil {
+		registry = make(map[string]ProviderModule)
+	}
+
+	shortName := p.GetName()
+
+	if _, exists := registry[shortName]; exists {
+		panic(fmt.Sprintf("provider %q already registered", shortName))
+	}
+
+	providers = append(providers,
+		hmc.Provider{
+			Name: ProviderPrefix + p.GetName(),
+		},
+	)
+
+	registry[shortName] = p
+}
+
+// List returns a copy of all registered providers
+func List() []hmc.Provider {
+	return slices.Clone(providers)
+}
+
+// CredentialPropagationFunc returns the credential propagation function for a given provider
+func CredentialPropagationFunc(fullName string) (
+	func(ctx context.Context, propnCfg *credspropagation.PropagationCfg, l logr.Logger) (enabled bool, err error), bool,
+) {
+	mu.RLock()
+	defer mu.RUnlock()
+
+	shortName := strings.TrimPrefix(fullName, ProviderPrefix)
+
+	module, ok := registry[shortName]
+	if !ok {
+		return nil, false
+	}
+
+	f := module.CredentialPropagationFunc()
+
+	return f, f != nil
+}
+
+// GetClusterGVK returns the GroupVersionKind for a provider's cluster resource
+func GetClusterGVK(shortName string) schema.GroupVersionKind {
+	mu.RLock()
+	defer mu.RUnlock()
+
+	module, ok := registry[shortName]
+	if !ok {
+		return schema.GroupVersionKind{}
+	}
+
+	return module.GetClusterGVK()
+}
+
+// GetClusterIdentityKinds returns the supported identity kinds for a given infrastructure provider
+func GetClusterIdentityKinds(infraName string) ([]string, bool) {
+	mu.RLock()
+	defer mu.RUnlock()
+
+	shortName := strings.TrimPrefix(infraName, InfraPrefix)
+
+	module, ok := registry[shortName]
+	if !ok {
+		return nil, false
+	}
+
+	list := slices.Clone(module.GetClusterIdentityKinds())
+
+	return list, list != nil
```

Comment:
better to check the length than the memory ref or even drop the second return-arg as in the GVK func signature

Reply:
2nd arg is needed to disable provider that is build-in, can change to to len(list) > 0